### DISCUSSION TITLE
Next.js doesn't respect Browserslist

### DISCUSSION THREAD

**Discussion by jaydenseric** [vercel/next.js#12826] [Help] [answered]
Out of the box, defining a [browserslist](https://github.com/browserslist/browserslist) query [the standard ways](https://github.com/browserslist/browserslist#config-file) in your project (e.g. via a `package.json` `browserslist` field) is not respected by Next.js for builds.

I see people hardcoding a query into the Next.js Babel preset config, like this:

https://github.com/zeit/next.js/blob/fa5be4971ad1a3fb8f6214d8f21cc03ca60d6457/test/integration/babel-custom/fixtures/targets-string/.babelrc#L7

But that is a poor solution as Browserslist won't find the query for other tools (e.g. ESLint plugins).

I'm not even sure what it does, since Next.js customises the `@babel/preset-env` `targets` for the server build, and maybe differently for 2 versions of the client build (modern?) but that's not documented anywhere. Does Next.js respect your query, but then exclude more targets on top, or does it totally ignore it?

Ideally we should be able to specify our Browserslist queries the standard ways, and Next.js will never add targets on top of it - only intersect that query and possibly reduce the number of targets as a result for the special builds.

**Comment by jaydenseric** [comment]
`@babel/preset-env` usually lets Browserslist retrieve the query if the `targets` option is not specified:

https://babeljs.io/docs/en/babel-preset-env#browserslist-integration

I tried this Babel config:

```json
{
  "presets": [
    [
      "next/babel",
      {
        "preset-env": {
          "targets": null
        }
      }
    ]
  ]
}
```

And it seems to cause the `package.json` `browserslist` field to be used, but I'm not sure if this will break the way Next.js makes the separate builds. I'm trying to read through the source code and it's not intuitive.

**Reply by itsjling** [reply]
This workaround no longer seems to work due to:

https://github.com/vercel/next.js/blob/497cac4b93de9ecf6c7ed79bd6557dcd3bb51be5/packages/next/build/babel/preset.ts#L89-L104

It would be great for a normal browserslist configuration in `package.json` or [any other officially supported method](https://github.com/browserslist/browserslist#queries) to be respected.

**Comment by xyy94813** [comment]
https://github.com/vercel/next.js/blob/f658b7641d5e500eca23b90ad3fa5705970f2525/packages/next/build/babel/preset.ts#L91-L97

> By default `@babel/preset-env` will use `browserslist` config sources unless either the targets or `ignoreBrowserslistConfig` options are set.

It seems we can use `browserslistrc`...

**Comment by feross** [comment]
Is there a more official solution to this issue yet?

Here's what we're doing, but it feels extremely brittle since Next.js can change how it uses `targets` or whether to obey or override that option at any time.

```js
const { readFileSync } = require('fs')
const { join } = require('path')

const browsersList = readFileSync(
  join(__dirname, '.browserslistrc'), 'utf8'
).split('\n')

// Next.js should use the .browserslistrc file
module.exports = {
  presets: [
    [
      "next/babel",
      {
        "preset-env": {
          targets: browsersList
        }
      }
    ]
  ]
}
```

**Comment by jaydenseric** [comment]
Guys, can you please all upvote this discussion by pressing the arrow at the top left of the discussion (on desktop):

<img width="218" alt="Screen Shot 2021-03-11 at 7 23 24 pm" src="https://user-images.githubusercontent.com/1754873/110757349-658bbe00-829f-11eb-9170-aabdfab11b2a.png">

Right now, 18 people have thumbed it but no one has actually upvoted the discussion to bring it to the attention of the wider community or Next.js team:

<img width="226" alt="Screen Shot 2021-03-11 at 7 24 37 pm" src="https://user-images.githubusercontent.com/1754873/110757524-9966e380-829f-11eb-9c99-7ef39a4a249b.png">

Also a side note, the GitHub design is a bit stupid I can only add a comment to this "discussion" by pressing a button called "Answer":

<img width="173" alt="Screen Shot 2021-03-11 at 7 29 48 pm" src="https://user-images.githubusercontent.com/1754873/110758042-2a3dbf00-82a0-11eb-84d6-70d14d63d97b.png">

**Comment by ghost** [comment]
I think the issue I've just filed might be related to this discussion: https://github.com/vercel/next.js/issues/23658. Next.js' autoprefixer setup is not adding the vendor prefixes you'd expect based on the browserslist in package.json (see issue for details).

**Comment by timbicker** [comment]
I think it is already possible like this in `babel.config.js`

```
const { NODE_ENV } = process.env;

const prod = NODE_ENV === "production";
const dev = NODE_ENV === "development";

module.exports = {
  "presets": [
    [
      "next/babel",
      {
        "preset-env": {
          targets: prod ? "defaults" : [
            "last 1 chrome version",
            "last 1 firefox version",
            "last 1 safari version"
          ]
        }
      }
    ]
  ]
}
```

Or am I missing something? See also the docs here: https://nextjs.org/docs/advanced-features/customizing-babel-config

I have not checked if it works because I did it just now, but I feel like the time to spin up the dev server is faster, so I think it worked.

**Reply by ghost** [reply]
That's already mentioned in the top comment. It's a workaround, rather than a fix to next.js not respecting the root browserslist config. As said in the top comment, with this workaround other tools won't pick up on the configuration.

**Comment by yangmingshan** [comment]
Seems now Next.js (Babel) will respect Browserslist.

**Reply by wereHamster** [reply]
Source? PR? Do you have an example?

**Reply by yangmingshan** [reply]
1. Initialize a project

```
yarn create next-app --typescript
```

2. Add `browserslist` to your `package.json`

```json
{
  "//": "...",
  "browserslist": [
    "last 2 Edge versions",
    "last 2 Opera versions",
    "last 2 Safari versions",
    "last 2 Chrome versions",
    "last 2 Firefox versions"
  ]
}
```

That's it.

And I looked at the source code, Next.js [will only add `targets` for the server or tests](https://github.com/vercel/next.js/blob/canary/packages/next/build/babel/preset.ts#L113), in other cases `targets` will be `undefined`, and Babel will look up for Browserslist config.

**Reply by lourd** [reply]
As of this writing, Next.js 12.0.7 does automatically use browserslist targets from config specified in a package.json file, but it still does not do so with the `.browerslistrc` file. You can see this by setting the debug option for preset-env `.babelrc`:

```
{
    "presets": [
      [
        "next/babel",
        {
          "preset-env": {
            "debug": true
          }
        }
      ]
    ],
  }
  ```

**Reply by pie6k** [reply]
@yangmingshan I think it still does not work (browserslist not respected). 

Also you mention `.bebelrc` file which switches from swc to babel compiler. With babel compiler it works, with swc (which is default now) it does not. 

Steps to reproduce:

```
yarn create next-app --typescript
```

in `_app.tsx` write

```ts
class FooSuper {
  async aa() {
    await "foo"
    console.log("foo super")
  }
}

console.log(FooSuper);
```

(or any 'modern' js code)

add to `package.json` 

```json
  "browserslist": [
    "Chrome 94"
  ]
```

run `yarn build`

Check build files for client and look for `foo super`

---

result: added _app code is bundled to something like this:

```js
e=n,(r=[{key:"aa",value:function(){return(n=o().mark((function n(){return o().wrap((function(n){for(;;)switch(n.prev=n.next){case 0:return n.next=2,"foo";case 2:console.log("foo super");case 3:case"end":return n.stop()}}),n)}))
```

expected result: I'd expect es6 class and native 'async/await' to be in client-side bundle as it is supported in defined `browserslist` query

**Reply by thexpand** [reply]
@pie6k I think it's because you're still using the SWC compiler, i.e. you have to opt out. The simplest way would be to add an empty `.babelrc` file.

**Reply by pie6k** [reply]
Thanks for pointing out - the point is I am aware of that and I'd prefer to use SWC for performance reasons.

**Reply by thexpand** [reply]
Fair enough. I was already having a custom `.babelrc` file with configuration for styled-components, as I'm waiting for a stable version of SWC plugin for styled-components. So, I had nothing to loose anyhow.

**Comment by trm217** [comment]
Does anyone have an idea on how this is handled with SWC? The browserslist key in package.json seems to have no effect and neither does having a custom config in .swcrc as metioned in the swc docs…

**Reply by rmales** [reply]
I was looking into this myself and the best answer I could find was https://github.com/vercel/next.js/issues/33985#issuecomment-1037227347, which basically seems to say Browserlist config isn't supported with SWC in Next for JS.

**Comment by timneutkens** [comment, marked as answer]
You can now add browserslist when enabling an experimental flag: https://github.com/vercel/next.js/discussions/33227#discussioncomment-3047837

**Comment by shay-alaluf** [comment]
Did someone find an actually solution? I have next 12.3 with swc with styled-jsx and somehow it works only on the latest versions of browsers (in all platform) and in other versions there are style breaks

**Comment by JodusNodus** [comment]
This is still an issue in next 15.

**Comment by krzkz94** [comment]
Can confirm, browserslist is completely ignored in NextJS 15 (15.5.2)

**Comment by tjazsilovsek** [comment]
What is the status here? How can we set it?
